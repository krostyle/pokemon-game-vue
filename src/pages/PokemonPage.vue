<template>
    <h1 v-if="!pokemon">Cargando... Espere Porfavor</h1>
    <div v-else>
        <h1>Â¿Quien es este Pokemon?</h1>
        <pokemon-picture :pokemonId="pokemon.id" :showPokemon="showPokemon"></pokemon-picture>
        <pokemon-options :pokemons="pokemonsArray" @selection="checkAnswer"></pokemon-options> 
        <template v-if="showAnswer">
            <h2>{{message}}</h2>    
            <button @click="newGame">Nuevo Juego</button>
        </template>
    </div>
</template>

<script>
import PokemonPicture from '@/components/PokemonPicture.vue'
import PokemonOptions from '@/components/PokemonOptions.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'


export default {
  components: { 
        PokemonPicture,
        PokemonOptions
      },
    name:'PokemonPage',
    data() {
        return {
            pokemonsArray: [],
            pokemon:null,
            showPokemon:false,
            showAnswer:false,
            message:''
        }
    },
    methods:{
        async mixPokemons(){
            this.pokemonsArray = await getPokemonOptions();
            const rndIndex = Math.floor(Math.random() * this.pokemonsArray.length);
            this.pokemon = this.pokemonsArray[rndIndex];
        },
        checkAnswer(id){
            this.showPokemon = true;
            this.showAnswer = true;
            if(id === this.pokemon.id){                
                this.message = `Correcto! es ${this.pokemon.name}`;
            }else{
                this.message = `Incorrecto! es ${this.pokemon.name}`;
            }
        },
        newGame(){
            this.pokemon=null;
            this.showAnswer = false;
            this.showPokemon = false;
            this.pokemonsArray = [];            
            this.mixPokemons();
        }
        
    },
    mounted(){
        this.mixPokemons();
    }
}
</script>

<style>

</style>