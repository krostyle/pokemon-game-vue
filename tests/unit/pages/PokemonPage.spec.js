import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'
import { pokemons } from "../mocks/pokemons.mocks";
describe('Pokemon Page', () => {
    let wrapper


    beforeEach(() => {
        const mixPokemonsArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemons')
        wrapper = shallowMount(PokemonPage)
    })

    test('should call mixPokemons when mounted', () => {
        const mixPokemonsArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemons')
        const wrapper = shallowMount(PokemonPage)
        expect(mixPokemonsArraySpy).toHaveBeenCalled();
    })

    test('must match the snapshot when loading the pokemon', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonsArray: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('should show PokemonPicture and PokemonOptions components', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonsArray: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const pokemonPicture = wrapper.find("pokemon-picture-stub")
        const pokemonOptions = wrapper.find("pokemon-options-stub")

        expect(pokemonPicture.exists()).toBeTruthy()
        expect(pokemonOptions.exists()).toBeTruthy()

        expect(pokemonPicture.attributes('pokemonid')).toBe(`${pokemons[0].id}`)
        expect(pokemonOptions.attributes('pokemons')).toBeTruthy()

    })

    test("test with checkAswer", async() => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonsArray: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(pokemons[0].id)
        const h2 = wrapper.find("h2")
        expect(h2.exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBeTruthy()
        expect(h2.text()).toBe(`Correcto! es ${pokemons[0].name}`)
    })
})