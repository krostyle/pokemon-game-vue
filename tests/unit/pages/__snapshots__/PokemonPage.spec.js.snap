// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon Page must match the snapshot when loading the pokemon 1`] = `
<div>
  <h1>Â¿Quien es este Pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
